lib/semantics/resolve-labels.cc

```diff
@@ -0,0 +1,969 @@
+template<typename A> constexpr bool IsFormat(const parser::Statement<A> &) {
+  return std::is_same_v<A, common::Indirection<parser::FormatStmt>>;
+}
+
+/// \brief Is this a legal branch target?
+/// Pattern match dependent on the standard we're enforcing
+template<typename A>
+constexpr bool IsLegalBranchTarget(const parser::Statement<A> &) {
+  return std::disjunction_v<std::is_same<A, parser::AssociateStmt>,
+      std::is_same<A, parser::EndAssociateStmt>,
+      std::is_same<A, parser::IfThenStmt>, std::is_same<A, parser::EndIfStmt>,
+      std::is_same<A, parser::SelectCaseStmt>,
+      std::is_same<A, parser::EndSelectStmt>,
+      std::is_same<A, parser::SelectRankStmt>,
+      std::is_same<A, parser::SelectTypeStmt>,
+      std::is_same<A, common::Indirection<parser::LabelDoStmt>>,
+      std::is_same<A, parser::NonLabelDoStmt>,
+      std::is_same<A, parser::EndDoStmt>,
+      std::is_same<A, common::Indirection<parser::EndDoStmt>>,
+      std::is_same<A, parser::BlockStmt>, std::is_same<A, parser::EndBlockStmt>,
+      std::is_same<A, parser::CriticalStmt>,
+      std::is_same<A, parser::EndCriticalStmt>,
+      std::is_same<A, parser::ForallConstructStmt>,
+      std::is_same<A, parser::ForallStmt>,
+      std::is_same<A, parser::WhereConstructStmt>,
+      std::is_same<A, parser::EndFunctionStmt>,
+      std::is_same<A, parser::EndMpSubprogramStmt>,
+      std::is_same<A, parser::EndProgramStmt>,
+      std::is_same<A, parser::EndSubroutineStmt>>;
+}
+template<>
+constexpr bool IsLegalBranchTarget(
+    const parser::Statement<parser::ActionStmt> &actionStmt) {
+  if (!isStrictF18) {
+    return true;
+  } else {
+    // XXX: do we care to flag these as errors? If we want strict F18, these
+    // statements should not even be present
+    return !(
+        std::holds_alternative<common::Indirection<parser::ArithmeticIfStmt>>(
+            actionStmt.statement.u) ||
+        std::holds_alternative<common::Indirection<parser::AssignStmt>>(
+            actionStmt.statement.u) ||
+        std::holds_alternative<common::Indirection<parser::AssignedGotoStmt>>(
+            actionStmt.statement.u) ||
+        std::holds_alternative<common::Indirection<parser::PauseStmt>>(
+            actionStmt.statement.u));
+  }
+}
+
+template<typename A>
+constexpr TargetStmtType ConsTrgtFlags(const parser::Statement<A> &statement) {
+  TargetStmtType targetStmtType{};
+  if (IsLegalDoTerm(statement)) {
+    targetStmtType.set(TargetStatementEnum::Do);
+  }
+  if (IsLegalBranchTarget(statement)) {
+    targetStmtType.set(TargetStatementEnum::Branch);
+  }
+  if (IsFormat(statement)) {
+    targetStmtType.set(TargetStatementEnum::Format);
+  }
+  return targetStmtType;
+}
+
+/// \brief \p opt1 and \p opt2 must be either present and identical or absent
+/// \param a  an optional construct-name (opening statement)
+/// \param b  an optional construct-name (ending statement)
+template<typename A> inline bool BothEqOrNone(const A &a, const A &b) {
```

Comment:
Why not just define this for `std::optional<Name>`? That seems to be the only way it is used.

Reply:
Why not let type inference do it?

Reply:
Abstraction over types is like charging a local readability tax in the hope that there's a *net* gain in readability &/or correctness when the abstraction is applied to multiple instantiations.  When there's but a single instantiation of a template and no reasonable expectation of others in the future, concrete types will probably be more readable.

Reply:
Concrete types were pushed everywhere possible